diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/SmartCastRule.kt

```diff
@@ -0,0 +1,173 @@
+        val thenBlock = then.findChildByType(BLOCK)
+
+        if (thenBlock != null) {
+            // Find all as expressions that are inside this current block
+            val asList = thenBlock.findAllNodesWithSpecificType(BINARY_WITH_TYPE).filter { it.text.contains(" as ")
+                    && it.findParentNodeWithSpecificType(BLOCK) == thenBlock }
+                    .filterNot { (it.getFirstChildWithType(REFERENCE_EXPRESSION)?.psi as KtNameReferenceExpression).hasLocalDeclaration() != null }
+            checkAsExpressions(asList, blocks)
+        } else {
+            val asList = then.findAllNodesWithSpecificType(BINARY_WITH_TYPE).filter {  it.text.contains(" as ") }
+            checkAsExpressions(asList, blocks)
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun checkAsExpressions(asList: List<ASTNode>, blocks: List<IsExpressions>) {
+        val asExpr = mutableListOf<AsExpressions>()
+
+        asList.forEach {
+            val split = it.text.split("as").map { part -> part.trim() }
+            asExpr.add(AsExpressions(split[0], split[1], it))
+        }
+
+        val exprToChange = asExpr.filter {
+            blocks.any { isExpr ->
+                isExpr.identifier == it.identifier
+                        && isExpr.type == it.type
+            }
+        }
+
+        if (exprToChange.isNotEmpty()) {
+            exprToChange.forEach {
+                SMART_CAST_NEEDED.warnAndFix(configRules, emitWarn, isFixMode, "${it.identifier} as ${it.type}", it.node.startOffset,
+                        it.node) {
+                    val dotExpr = it.node.findParentNodeWithSpecificType(DOT_QUALIFIED_EXPRESSION)!!
+                    val afterDotPart = dotExpr.text.split(".")[1]
+                    val text = "${it.identifier}.$afterDotPart"
+                    dotExpr.treeParent.addChild(KotlinParser().createNode(text), dotExpr)
+                    dotExpr.treeParent.removeChild(dotExpr)
+                }
+            }
+        }
+    }
+
+    @Suppress("UnsafeCallOnNullableType")
+    private fun handleWhenCondition(node: ASTNode) {
+        /*
+            Check if there is WHEN_CONDITION_IS_PATTERN. If so delete 'as' in it's block
+            or call expression if it doesn't have block
+         */
+
+        val identifier = node.getFirstChildWithType(REFERENCE_EXPRESSION)?.text
+
+        node.getAllChildrenWithType(WHEN_ENTRY).forEach {
+            if (it.hasChildOfType(WHEN_CONDITION_IS_PATTERN) && identifier != null) {
+                val type = it.getFirstChildWithType(WHEN_CONDITION_IS_PATTERN)!!
+                        .getFirstChildWithType(TYPE_REFERENCE)!!.text
+
+                val callExpr = it.getFirstChildWithType(CALL_EXPRESSION)!!
+                val blocks = listOf(IsExpressions(identifier, type))
+
+                handleThenBlock(callExpr, blocks)
+            }
+        }
+    }
+
+    private fun KtNameReferenceExpression.hasLocalDeclaration(): KtProperty? = parents
```

Comment:
looks like you search usage of variables. I suggest to use Petr's mechanism introduced in https://github.com/cqfn/diKTat/pull/352